File: app/src/main/java/me/tagavari/airmessage/redux/ReduxEventConnection.java
package me.tagavari.airmessage.redux;

import me.tagavari.airmessage.enums.ConnectionErrorCode;
import me.tagavari.airmessage.enums.ConnectionState;

public abstract class ReduxEventConnection {
	@ConnectionState
	public abstract int getState();
	
	public static class Connected extends ReduxEventConnection {
		@Override
		public int getState() {
			return ConnectionState.connected;
		}
	}
	
	public static class Connecting extends ReduxEventConnection {
		@Override
		public int getState() {
			return ConnectionState.connecting;
		}
	}
	
	public static class Disconnected extends ReduxEventConnection {
		@ConnectionErrorCode int code;
		
		public Disconnected(@ConnectionErrorCode int code) {
			this.code = code;
		}
		
		@Override
		public int getState() {
			return ConnectionState.disconnected;
		}
		
		@ConnectionErrorCode
		public int getCode() {
			return code;
		}
	}
}

File: app/src/main/java/me/tagavari/airmessage/redux/ReduxEventConnection.kt
package me.tagavari.airmessage.redux

import me.tagavari.airmessage.enums.ConnectionErrorCode
import me.tagavari.airmessage.enums.ConnectionState

abstract class ReduxEventConnection {
	@get:ConnectionState
	abstract val state: Int
	
	class Connected : ReduxEventConnection() {
		override val state = ConnectionState.connected
	}
	
	class Connecting : ReduxEventConnection() {
		override val state = ConnectionState.connecting
	}
	
	class Disconnected(
		@field:ConnectionErrorCode @get:ConnectionErrorCode
		@param:ConnectionErrorCode val code: Int
	) : ReduxEventConnection() {
		override val state = ConnectionState.connected
	}
}